src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,55 +27,116 @@ class DictionarySegment : public BaseSegment {
+      const auto insert_index = std::distance(_dictionary->begin(), insert_iter);
+      temp_attribute_vector.emplace_back(insert_index);
+    }
+
+    const auto local_unique_values_count = unique_values_count();
+
+    if (local_unique_values_count <= std::numeric_limits<uint8_t>::max()) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint8_t>>(temp_attribute_vector);
+    } else if (local_unique_values_count <= std::numeric_limits<uint16_t>::max()) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint16_t>>(temp_attribute_vector);
+    } else if (local_unique_values_count <= std::numeric_limits<uint32_t>::max()) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint32_t>>(temp_attribute_vector);
+    } else {
+      throw std::runtime_error("The segment contains more than UINT32_MAX distinct values.");
+    }
+  }
 
   // SEMINAR INFORMATION: Since most of these methods depend on the template parameter, you will have to implement
   // the DictionarySegment in this file. Replace the method signatures with actual implementations.
 
   // return the value at a certain position. If you want to write efficient operators, back off!
-  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override;
+
+  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override { return get(chunk_offset); }
 
   // return the value at a certain position.
-  T get(const size_t chunk_offset) const;
+  T get(const size_t chunk_offset) const {
+    const auto value_id = _attribute_vector->get(chunk_offset);
+    return value_by_value_id(value_id);
+  }
 
   // dictionary segments are immutable
-  void append(const AllTypeVariant& val) override;
+  void append(const AllTypeVariant& val) override {
+    throw std::runtime_error("Appending to an immutable dictionary segment is not allowed.");
+  }
 
   // returns an underlying dictionary
-  std::shared_ptr<const std::vector<T>> dictionary() const;
+  std::shared_ptr<const std::vector<T>> dictionary() const { return _dictionary; }
 
   // returns an underlying data structure
-  std::shared_ptr<const BaseAttributeVector> attribute_vector() const;
+  std::shared_ptr<const BaseAttributeVector> attribute_vector() const { return _attribute_vector; }
 
   // return the value represented by a given ValueID
-  const T& value_by_value_id(ValueID value_id) const;
+
+  const T& value_by_value_id(ValueID value_id) const { return _dictionary->at(value_id); }
 
   // returns the first value ID that refers to a value >= the search value
   // returns INVALID_VALUE_ID if all values are smaller than the search value
-  ValueID lower_bound(T value) const;
+  ValueID lower_bound(T value) const {
+    const auto iterator = std::lower_bound(_dictionary->begin(), _dictionary->end(), value);
+    return (iterator != _dictionary->end())
+               ? ValueID{static_cast<ValueID::base_type>(std::distance(_dictionary->begin(), iterator))}
```

Comment:
It should be possible to leave out the `static_cast` here and just use the `ValueID{}` constructor.
```suggestion
               ? ValueID{std::distance(_dictionary->begin(), iterator)}
```

Reply:
This does not work for us because gcc/x86_64-linux-gnu/9 throws a compiler error here: `narrowing conversion [...] std::distance[...] {aka ‘long int’} to ‘uint32_t’ {aka ‘unsigned int’}`. Do you know how to resolve this but with cleaner code like you suggested?